formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonPath.kt

```diff
@@ -0,0 +1,81 @@
+package kotlinx.serialization.json.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+
+internal class JsonPath {
+    /*
+     * Serial descriptor OR map key.
+     * Map key can also implement SD, so it also has '-2' as current index
```

Comment:
`Map key can also implement SD` - how? Serial descriptor is not `@Serializable`

Reply:
But serializable class that implements it is :)

This is a prototype, so there is no tests for that, but it may be the case anyway